crates/ruff_linter/src/noqa.rs

```diff
@@ -524,26 +547,140 @@ fn add_noqa_inner(
     let directives = NoqaDirectives::from_commented_ranges(comment_ranges, path, locator);
 
+    let comments =
+        find_noqa_comments_by_line(diagnostics, locator, &exemption, &directives, noqa_line_for);
+
+    let edits = build_noqa_edits_by_line(
+        comments
+            .into_iter()
+            .flatten()
+            .sorted_by_key(|comment| comment.line)
+            .group_by(|comment| comment.line)
+            .into_iter()
+            .map(|(line, comments)| (line, comments.collect()))
+            .collect(),
+        locator,
+        line_ending,
+    );
+
+    let contents = locator.contents();
+
+    let mut output = String::with_capacity(contents.len());
+    let mut last_append = TextSize::default();
+
+    for (_, edit) in edits {
+        output.push_str(&contents[TextRange::new(last_append, edit.start())]);
+
+        if let Some(content) = edit.content() {
+            if content != &contents[edit.range()] {
+                count += 1;
+            }
+            output.push_str(content);
+        }
+
+        last_append = edit.end();
+    }
+
+    output.push_str(&contents[TextRange::new(last_append, TextSize::of(contents))]);
+
+    (count, output)
+}
+
+fn build_noqa_edits_by_diagnostic(
+    comments: Vec<Option<NoqaComment>>,
+    diagnostics: &[Diagnostic],
+    locator: &Locator,
+    line_ending: LineEnding,
+) -> Vec<Option<Edit>> {
+    let mut edits = Vec::default();
+    for (i, comment) in comments.into_iter().enumerate() {
+        match comment {
+            Some(comment) => {
+                debug_assert_eq!(comment.diagnostic, &diagnostics[i]);
+                let line = locator.full_line(comment.line);
+                if let Some(edit) = generate_noqa_edit(
+                    comment.directive,
+                    comment.line,
+                    RuleSet::from_rule(comment.diagnostic.kind.rule()),
+                    line,
+                    locator.line_start(comment.line),
+                    line_ending,
+                ) {
+                    edits.push(Some(edit));
+                }
+            }
+            None => edits.push(None),
+        }
+    }
+    edits
+}
+
+fn build_noqa_edits_by_line(
+    comments_by_line: BTreeMap<TextSize, Vec<NoqaComment>>,
+    locator: &Locator,
+    line_ending: LineEnding,
+) -> BTreeMap<TextSize, Edit> {
+    let mut edits = BTreeMap::default();
+    for (offset, matches) in comments_by_line {
+        if matches.is_empty() {
+            continue;
+        }
+        let line = locator.full_line(offset);
+        let directive = matches.first().unwrap().directive.clone();
+        let rules: Vec<_> = matches
+            .into_iter()
+            .map(|NoqaComment { diagnostic, .. }| diagnostic.kind.rule())
+            .collect();
+        if let Some(edit) = generate_noqa_edit(
+            directive,
+            offset,
+            RuleSet::from_rules(rules.as_slice()),
+            line,
```

Comment:
Nit: `RuleSet` implements `FromIter`
```suggestion
        let rules: RuleSet = matches
            .into_iter()
            .map(|NoqaComment { diagnostic, .. }| diagnostic.kind.rule())
            .collect();
```